discovery: add support for Managed Identity authentication in Azure SD

marratj opened the pull request:
Azure supports Managed Identity authentication for a while now: https://docs.microsoft.com/en-us/azure/active-directory/managed-identities-azure-resources/overview

This PR is to add support in the Azure SD for Managed Identity auth as an alternative to OAuth with an Azure AD Service Principal.

simonpasquier commented on discovery/azure/azure.go:
```diff
@@ -170,13 +172,32 @@ func createAzureClient(cfg SDConfig) (azureClient, error) {
 
 	var c azureClient
+
 	oauthConfig, err := adal.NewOAuthConfig(activeDirectoryEndpoint, cfg.TenantID)
```
shouldn't it be moved to the "OAuth" switch case too?

marratj replied:
You're right, just changed it.

simonpasquier commented on discovery/azure/azure.go:
```diff
@@ -170,13 +172,32 @@ func createAzureClient(cfg SDConfig) (azureClient, error) {
+			return azureClient{}, err
+		}
+	default:
+		return azureClient{}, fmt.Errorf("Unsupported authentication_type set. Needs to be 'OAuth' or 'ManagedServiceIdentity'")
```
This should be handled by `UnmarshalYAML()` at L86.

marratj replied:
Right, as the SubscriptionID check is also handled there, understood.

I have put the check into UnmarshalYAML() and removed the default case.

Hope this looks ok now :)

marratj commented on discovery/azure/azure.go:
```diff
@@ -91,6 +93,11 @@ func (c *SDConfig) UnmarshalYAML(unmarshal func(interface{}) error) error {
 	}
+
+	if c.AuthenticationMethod != "OAuth" && c.AuthenticationMethod != "ManagedServiceIdentity" {
+		return fmt.Errorf("Unsupported authentication_type set. Needs to be 'OAuth' or 'ManagedServiceIdentity'")
```
Ok, I have changed it so it displays the invalid value and shows the available choices.

However, on writing a test on the check I'm honestly a bit lost. I actually don't have any real experience in writing unit tests (this is one of my first PRs). I tried to look at how other SD plugins test for this stuff, but I see that many don't even have tests (like ec2, dns & gce).

For any hints on how to structure that test I would be grateful :-)

marratj replied:
OK, so I added also a bad config test similar to the OpenStack SD one.

marratj commented on docs/configuration/configuration.md:
```diff
@@ -274,13 +274,16 @@ See below for the configuration options for Azure discovery:
+# The subscription ID. Always required.
 subscription_id: <string>
-# The tenant ID.
+# The tenant ID. Only required with authentication_method OAuth.
```
It's a good idea, but I agree that we might want to have it as a separate PR.

simonpasquier commented:
:+1:

simonpasquier commented:
@brian-brazil can you have a look at this?

marratj commented:
Any chance it can be merged in time for 2.5? Or is anything missing? :-)

simonpasquier commented on discovery/azure/azure_test.go:
```diff
@@ -20,6 +20,12 @@ import (
 	"github.com/Azure/azure-sdk-for-go/arm/compute"
 )
 
+func TestAuthenticationMethodIsValid(t *testing.T) {
```
This test doesn't bring much, I'd remove it.

simonpasquier commented on docs/configuration/configuration.md:
```diff
@@ -274,13 +274,16 @@ See below for the configuration options for Azure discovery:
 # The Azure environment.
 [ environment: <string> | default = AzurePublicCloud ]
-# The subscription ID.
+# The authentication method
```
I'd add a blank line as the following parameters relate to authentication only.

marratj commented on discovery/azure/azure.go:
```diff
@@ -63,21 +63,23 @@ var (
+		Port:                 80,
+		RefreshInterval:      model.Duration(5 * time.Minute),
+		Environment:          azure.PublicCloud.Name,
+		AuthenticationMethod: "OAuth",
```
I have defined as constants (I hope the place I put them in is ok).

Also, I have changed the name of the Auth Method from ManagedServiceIdentity to ManagedIdentity to reflect the fact that Microsoft has renamed it.

marratj commented:
@simonpasquier @brian-brazil I think I have addressed all your comments, can you please have another look? :)

Also, I have changed the name of the Auth Method from ManagedServiceIdentity to ManagedIdentity to reflect the fact that Microsoft has renamed it.

marratj commented:
rebased again after #4907 

@simonpasquier any chance we get this PR merged soon, as it is already 3 months old and was approved twice by you already?

marratj commented:
@brian-brazil any objections against merging this PR?

brian-brazil commented on docs/configuration/configuration.md:
```diff
@@ -275,13 +275,17 @@ See below for the configuration options for Azure discovery:
 subscription_id: <string>
-# The tenant ID.
+# The tenant ID. Only required with authentication_method OAuth.
 tenant_id: <string>
```
This should be marked optional. And below.

marratj replied:
They are kinda semi-optional, as they are only optional when using ManagedIdentity (in which case they are not even used).

However, they are required when using OAuth as authentication method (which is the default) and Prometheus will fail hard when they are not set in this case (introduced with #4907 )

brian-brazil replied:
They're only required if they're always required.

marratj replied:
OK, I have marked them now as optional in the docs.

marratj commented:
@brian-brazil done

brian-brazil commented:
Thanks!
